java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java

```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
-        //    and check for a Package property. If found, prepend the property
-        //    value to the classname.
-        // 3. If that fails, check for an Default.Package property. If found,
-        //    prepend the property value to the classname.
+        // 1. Convert the Slice type ID into a classname (e.g., ::M::X becomes M.X).
+        // 2. Check if the application has defined any package prefixes for the top-level module (e.gm "M").
+        //    Attempt to resolve the <package-prefix>+<class-name> in order trying each defined package
+        //    prefix.
+        // 3. If the above step fails, search for a Default.Package property. If found,
+        //    prepend its value to the classname. Otherwise, attempt to use the
+        //    classname directly.
         //
         String fullyQualifiedClassName;
-        boolean addClass = false;
 
         //
         // See if we've already translated this type ID before.
         //
-        synchronized(this)
+        synchronized (this)
         {
             fullyQualifiedClassName = _sliceTypeIdToClassMap.get(typeId);
         }
 
         //
-        // It's a new type ID, so first convert it into a Java class name.
+        // The fully qualified class name was already resolved, keep using it.
         //
-        if(fullyQualifiedClassName == null)
+        if (fullyQualifiedClassName != null)
         {
-            fullyQualifiedClassName = com.zeroc.Ice.Util.typeIdToClass(typeId);
-            addClass = true;
+            return getConcreteClass(fullyQualifiedClassName);
         }
 
         //
-        // See if we can find the class without any prefix.
+        // It's a new type ID, so first convert it into a Java class name.
         //
-        c = getConcreteClass(fullyQualifiedClassName);
+        fullyQualifiedClassName = com.zeroc.Ice.Util.typeIdToClass(typeId);
 
         //
-        // See if the application defined an Ice.Package.MODULE property.
+        // See if the application defined any package prefixes with Ice.Package.MODULE property.
         //
-        if(c == null)
+        int pos = typeId.indexOf(':', 2);
+        if (pos != -1)
         {
-            int pos = typeId.indexOf(':', 2);
-            if(pos != -1)
+            String topLevelModule = typeId.substring(2, pos);
+            String[] packagePrefixes = _builtInModulePackagePrefixes.get(topLevelModule);
+            if (packagePrefixes == null)
             {
-                String topLevelModule = typeId.substring(2, pos);
-                String[] packagePrefixes = _builtInModulePackagePrefixes.get(topLevelModule);
-                if (packagePrefixes == null)
-                {
-                    packagePrefixes = _initData.properties.getPropertyAsListWithDefault("Ice.Package." + topLevelModule, null);
-                }
+                packagePrefixes = _initData.properties.getPropertyAsListWithDefault("Ice.Package." + topLevelModule, null);
```

Comment:
Why WithDefault? The implicit default is presumably null.

Reply:
You are right, getPropertyAsList defaults to null.